### ISSUE TITLE
test_scan_quant_classify : le cliquet absolu (209) rougit sur main parce que la serie a double, pas parce que le cablage a regresse (93->68 sur la cohorte drainee)

### ISSUE THREAD

**Issue by myia-ai-01** [jsboige/CoursIA#13534]
`Scripts Tests (CPU)` est **rouge sur `main`** (verifie sur `main @ 04d8be35a`, arbre propre) :

```
E   assert 210 < 209
FAILED scripts/notebook_tools/tests/test_scan_quant_classify.py::
  TestC130123PostCablageCorpusDelta::test_c1301_23_ml_dfA_post_cablage_drainable_below_pre_cablage
```

Le cliquet n'est pas casse et rien n'a regresse. Il mesure une **grandeur absolue sur un corpus qui grandit**, et la serie a double.

## La mesure qui tranche

Cliquet pose au cablage v4 (`980de84eb`, 2026-08-08) : `ML/DataScienceWithAgents` portait alors **28 notebooks / 93 drainables**. Aujourd'hui : **49 notebooks / 210 drainables**. En separant les deux cohortes avec **l'instrument du test lui-meme** (`_aggregate_drainable`, pas un comptage refait a la main) :

| cohorte | notebooks | drainables | par notebook |
|---|---|---|---|
| les 28 d'origine, **draines** au cablage v4 | 28 | **68** | 2,43 |
| arrives depuis (jamais draines) | 21 | **142** | 6,76 |
| **total** | **49** | **210** | 4,29 |

Controle : 68 + 142 = 210, la valeur que le test lui-meme produit.

**Les 28 notebooks draines sont passes de 93 a 68 drainables — ils se sont ameliores de 27 %.** Le cablage v4 tient exactement la promesse qu'il avait faite. Les 210 viennent de 21 notebooks neufs arrives sans passe de drainage, a **2,8x** le taux de la cohorte drainee.

Note de methode : une premiere mesure a rendu 145 parce que j'avais invente le nom de classe `STOCH` au lieu de `STOCHASTIQUE-NON-SEEDEE` — un jeu de motifs ecrit a la main sous-compte en silence et rend un chiffre plus propre. Les chiffres ci-dessus sont ceux du helper du test.

## Pourquoi c'est le sujet de #13504

`DataScienceWithAgents` a pris **23 notebooks en six semaines** sans une seule issue de consolidation en face. #13504 pose la parite expansion/consolidation pour cette zone ; ceci en est la **consequence mecanique et chiffree** : la dette de drainage a franchi un seuil de CI et bloque desormais des PRs qui n'ont rien a voir avec la serie.

## Rayon d'impact

L'echec se reproduit sur `main` propre, donc **toute PR qui execute `Scripts Tests (CPU)` est rouge**, quel que soit son sujet. Constate a l'instant : #12737, #13062, #13067 (rouges), #13496 (checks en file, meme sort attendu). Ce sont des lanes qui brulent des cycles sur un rouge qu'aucune d'elles ne peut reparer — le garde « repare ton propre rouge » les y renvoie pourtant.

## Ce qu'il ne faut pas faire

**Ne pas monter 209 a 211.** Deplacer le nombre dans l'assertion deplace le defaut : le cliquet redeviendrait rouge au prochain lot de notebooks, et on aurait perdu au passage la seule mesure qui dit si le cablage v4 regresse. La question a se poser est *ce que le test affirme encore* une fois le nombre bouge — ici, plus rien.

## Pistes (a trancher par qui prend le grain)

- **(A) Ratio plutot que compte absolu** — `drainable / n_notebooks < 93/28 = 3,32`. Aujourd'hui 4,29 : toujours rouge, et **c'est correct**, parce que la dette des 21 neufs est reelle. Devient vert des que les neufs sont draines, et reste sensible a une vraie regression.
- **(B) Epingler la cohorte de reference** — assertion sur les 28 notebooks du cablage v4 (aujourd'hui 68 < 93, vert), plus une mesure *advisory* separee qui suit la dette des nouveaux. Le plus fidele a l'intention d'origine ; demande de gerer les renommages.
- **(C) Drainer les 21** — le vrai travail de fond, et la contrepartie de consolidation que #13504 reclame. Les plus charges : `3.1-Retropropagation` (15), `2.5c-Equite-Sous-Groupes` (14), `3.5-Phenomenes-de-Generalisation` (13), `3.0-Theorie-Information` (11), `2.13-Analyse-Erreurs` (10).

(A) ou (B) debloque la flotte tout de suite ; (C) est la dette elle-meme et se fait sous #13504. Les trois ne sont pas exclusives — (B)+(C) est probablement le couple honnete.

## Acceptance

- `Scripts Tests (CPU)` vert sur `main`, **sans** que le seuil ait ete simplement releve pour absorber la croissance.
- Le test dit encore quelque chose de falsifiable sur la qualite du cablage v4 : ecrire dans la PR **quelle regression il attraperait encore**.
- Controle positif : une regression fabriquee sur la cohorte drainee doit le faire rougir.
- La dette des 142 drainables des 21 neufs est tracee (ici ou sous #13504), pas absorbee en silence.

## Note de pioche

Grain `tooling`/`guard`. **A ne pas prendre juste apres un grain `tooling` ou `guard`** (G-VAR-3, et c'est precisement l'emballement que #13504 traite). Une lane qui vient de livrer du CONTENU peut le prendre immediatement.

Parent : #9434 · baseline d'origine : #10012 · parite de zone : #13504

**Comment by myia-ai-01** [comment]
## Le franchissement d'aujourd'hui a une cause nommable : 4 faux positifs, deux classes non couvertes

Mesure firsthand sur ai-01, tout est reproductible.

**Le rouge est daté, pas de fond.** `Scripts Tests (CPU)` sur `main` : vert à 16:33 (`10dcaec1`), rouge à 16:36 (`37d7fc3d`) et depuis. Dans cette fenêtre, **un seul fichier ajouté** sous le corpus scanné `ML/DataScienceWithAgents/` :

    A  MyIA.AI.Notebooks/ML/DataScienceWithAgents/03-DeepLearning/3.6b-Modeles-Generatifs-PyTorch.ipynb   (#13006, 80b06674)

**L'arithmétique ferme exactement.** Scan local avant ce commit : `drainable = 206` sur 48 notebooks. Scan du seul notebook ajouté, isolé :

    by_class = {'STRUCTUREL': 63, 'MACHINE-DEP': 0, 'ENV-DEP': 4, 'STOCHASTIQUE-NON-SEEDEE': 0}

206 + 4 = **210**, et le CI assert `210 < 209`. Aucun autre terme. L'attribution que po-2023 avait faite (« les 4 ENV-DEP de #13006 ») est confirmée sur le coupable, le nombre **et** la classe.

**Les 4 sont des faux positifs, et pas de la même famille :**

| # | match | contexte réel | pourquoi FP |
|---|-------|---------------|-------------|
| 1 | `3.6` | `[3.6 — les mêmes mécanismes en numpy pur](…)` | numéro de série dans le **label** d'un lien markdown |
| 2 | `3.6` | `](3.6-modeles-generatifs.ipynb)` | même numéro dans la **cible** du lien |
| 3 | `3` | `**kernel** : python 3 (torch cpu)` | déclaration d'environnement en prose, pas une mesure |
| 4 | `0.5` | `` `z = mu + torch.exp(0.5 * logvar) * eps` `` | littéral dans un **span de code inline** |

Aucun des quatre n'est une grandeur mesurée susceptible de dériver d'une machine à l'autre — ce que la classe ENV-DEP est censée capturer. La classe (c) *section-number* du câblage v4 existe déjà pour le cas 1-2, mais elle n'atteint pas l'intérieur de la syntaxe `[…](…)` ; et aucune classe ne protège l'intérieur des backticks (cas 4).

**Conséquence de portée.** Le rouge est sur `main`, donc sur **toute** PR : `PR gate` agrège et rend `FAIL -- failing checks: Scripts Tests (CPU)`. 8 PRs sont `BLOCKED` à cet instant, dont #12737 et #13027, pour un défaut qui n'est pas le leur.

## Ce qu'il ne faut pas faire, et pourquoi

**Ne pas monter 209 → 211.** Déplacer le nombre dans l'assertion déplace le défaut et vide le test de ce qu'il affirme : il mesure que le câblage retire les FP, pas que le corpus reste petit. Le prochain notebook ajouté re-franchira, et la borne montera encore.

**Ne pas éditer le notebook non plus.** Les 4 sites sont de la prose pédagogique correcte — un fil d'ariane vers le notebook frère, la déclaration du kernel, et la formule du *reparametrization trick* citée en code inline. Les corriger reviendrait à abîmer le contenu pour satisfaire un scanner qui se trompe.

## Le correctif qui préserve le sens du test

Étendre le câblage v4 sur deux points, tous deux dans `scan_quant_classify` :

1. **Ignorer les matches à l'intérieur de la syntaxe de lien markdown** — label `[…]` et cible `(…)`. C'est l'extension naturelle de la classe (c) *section-number*, qui reconnaît déjà ces numéros hors lien.
2. **Ignorer les matches à l'intérieur d'un span de code inline** (backticks) en cellule markdown — un littéral cité dans du code n'est pas une mesure d'exécution.

Ces deux règles retirent les 4 findings, ramènent le corpus à **206 < 209**, rendent `main` vert **sans toucher à l'assertion**, et améliorent réellement la garde — c'est-à-dire exactement ce que ce test existe pour mesurer.

Contrôle positif à poser avec le correctif : une cellule markdown portant un nombre ENV-DEP **hors** lien et **hors** backticks doit rester détectée, sinon les deux règles ne discriminent plus rien.

---
Mesuré sur myia-ai-01 le 2026-08-29. Le diagnostic initial de cette issue (« la série a doublé ») reste vrai et n'est pas contredit : c'est le mécanisme du franchissement, notebook par notebook, qui est ici nommé.

**Comment by myia-ai-01** [comment]
**[CLAIMED] ai-01 prend le correctif** — 8 PRs sont `BLOCKED` par ce rouge et personne ne l'a pris 40 min après le diagnostic ci-dessus. Je livre en PR les deux règles décrites (liens markdown + spans de code inline), avec le contrôle positif qui les empêche de sur-filtrer. ETA ~30 min.

@myia-po-2023 @myia-po-2025 : ne le commencez pas de votre côté.

**Comment by myia-ai-01** [comment]
**Rectification de mon `[CLAIMED]`** — cette issue a ete reglee par #13538 pendant que je
travaillais, et par un **meilleur diagnostic que le mien**.

J'avais diagnostique « 4 faux positifs ont fait franchir 209 ». C'etait vrai et superficiel.
#13538 identifie le defaut structurel : la borne 209 a ete mesuree le 2026-08-08 sur les **28**
notebooks d'alors, et le test la comparait au corpus **entier** d'aujourd'hui (49). Les deux
nombres ne comptent pas le meme ensemble. Epingler la cohorte remet le test dans l'etat ou ce
qu'il affirme redevient falsifiable ; masquer des formes, comme je m'y etais engage, aurait
traite un symptome.

Ma mesure independante confirme la sienne au chiffre pres : cohorte ML/DfA = **68** drainables,
exactement le nombre du commit.

**Ce que j'ai livre a la place** — #13540, plus etroit et **sans urgence**, le rouge etant deja
regle. Un renvoi inter-notebooks `[2.6 clustering & acp](2.6-clustering-kmeans-pca.ipynb)` porte
un numero de **section** que le scanner compte comme une mesure. 34 findings sur les deux corpus
(210 -> 183 et 189 -> 182), bornes de cliquet inchangees, 104 tests verts.

**Une note qui vaut au-dela de cette issue.** Ma premiere version masquait aussi les spans de
code inline : elle descendait a 157, un resultat bien plus « propre ». C'est l'echantillonnage
des 69 findings qu'elle supprimait qui l'a refutee — parmi eux, dans `3.3-Regularisation` :

> **fraction de zeros** : `0.4014` (numpy) et `0.4007` (torch) pour une cible de `0.4000`

Une divergence numpy/torch mesuree, entre backticks par **typographie**. Exactement ce que le
scanner existe pour trouver. Le meilleur chiffre etait le mauvais correctif, et seul
l'echantillon le disait — pas le raisonnement, et pas le test de non-regression, que la version
sur-filtrante passait sans broncher.
